File: android-agent/app/src/main/java/com/androidscript/agent/runtime/Interpreter.kt
package com.androidscript.agent.runtime

import android.util.Log

/**
 * Interpreter for AndroidScript
 * Executes Abstract Syntax Tree (AST) using visitor pattern
 */
class Interpreter : ASTVisitor {

    private val global: Environment = Environment()
    private var environment: Environment = global
    private var lastValue: Value = NilValue
    private val errors = mutableListOf<String>()

    companion object {
        private const val TAG = "Interpreter"
    }

    /**
     * Get the global environment (for registering built-in functions)
     */
    fun getGlobalEnvironment(): Environment = global

    /**
     * Get execution errors
     */
    fun getErrors(): List<String> = errors
    fun hasErrors(): Boolean = errors.isNotEmpty()

    /**
     * Execute a list of statements
     */
    fun execute(statements: List<Statement>) {
        for (stmt in statements) {
            try {
                execute(stmt)
            } catch (e: ReturnException) {
                reportError("Return statement outside of function")
            } catch (e: BreakException) {
                reportError("Break statement outside of loop")
            } catch (e: ContinueException) {
                reportError("Continue statement outside of loop")
            } catch (e: Exception) {
                reportError("Runtime error: ${e.message}")
                Log.e(TAG, "Runtime error", e)
            }
        }
    }

    /**
     * Execute a single statement
     */
    fun execute(stmt: Statement) {
        stmt.accept(this)
    }

    /**
     * Evaluate an expression
     */
    fun evaluate(expr: Expression): Value {
        expr.accept(this)
        return lastValue
    }

    /**
     * Execute a block of statements with a new environment
     */
    private fun executeBlock(statements: List<Statement>, env: Environment) {
        val previous = environment
        try {
            environment = env
            for (stmt in statements) {
                execute(stmt)
            }
            environment = previous
        } catch (e: Exception) {
            environment = previous
            throw e
        }
    }

    /**
     * Call a function (native or user-defined)
     */
    private fun callFunction(callee: Value, args: List<Value>): Value {
        // Native function
        if (callee is NativeFunctionValue) {
            return callee.function(args)
        }

        // User-defined function
        if (callee is FunctionValue) {
            val func = callee.function

            // Check argument count
            if (args.size != func.parameters.size) {
                throw RuntimeException(
                    "Expected ${func.parameters.size} arguments but got ${args.size}"
                )
            }

            // Create new environment for function execution
            val funcEnv = Environment(func.closure)

            // Bind parameters
            for (i in args.indices) {
                funcEnv.define(func.parameters[i], args[i])
            }

            // Execute function body
            return try {
                val previous = environment
                environment = funcEnv
                execute(func.body)
                environment = previous
                NilValue  // No explicit return
            } catch (e: ReturnException) {
                e.value
            }
        }

        throw RuntimeException("Value is not callable")
    }

    /**
     * Report runtime error
     */
    private fun reportError(message: String) {
        errors.add(message)
        Log.e(TAG, message)
    }

    // ========================================================================
    // Expression Visitors
    // ========================================================================

    override fun visit(expr: BinaryExpr) {
        val left = evaluate(expr.left)
        val right = evaluate(expr.right)

        lastValue = when (expr.op.type) {
            TokenType.PLUS -> left + right
            TokenType.MINUS -> left - right
            TokenType.MULTIPLY -> left * right
            TokenType.DIVIDE -> left / right
            TokenType.MODULO -> left % right
            TokenType.EQUAL -> BoolValue(left.equals(right))
            TokenType.NOT_EQUAL -> BoolValue(left.notEquals(right))
            TokenType.LESS -> BoolValue(left.lessThan(right))
            TokenType.LESS_EQUAL -> BoolValue(left.lessEqual(right))
            TokenType.GREATER -> BoolValue(left.greaterThan(right))
            TokenType.GREATER_EQUAL -> BoolValue(left.greaterEqual(right))
            TokenType.LOGICAL_AND -> BoolValue(left.isTruthy() && right.isTruthy())
            TokenType.LOGICAL_OR -> BoolValue(left.isTruthy() || right.isTruthy())
            else -> throw RuntimeException("Unknown binary operator: ${expr.op.type}")
        }
    }

    override fun visit(expr: UnaryExpr) {
        val operand = evaluate(expr.operand)

        lastValue = when (expr.op.type) {
            TokenType.MINUS -> -operand
            TokenType.LOGICAL_NOT -> !operand
            else -> throw RuntimeException("Unknown unary operator: ${expr.op.type}")
        }
    }

    override fun visit(expr: LiteralExpr) {
        lastValue = when (expr.value.type) {
            TokenType.TRUE -> BoolValue(true)
            TokenType.FALSE -> BoolValue(false)
            TokenType.NULLPTR -> NilValue
            TokenType.INTEGER -> IntValue(expr.value.intValue)
            TokenType.FLOAT -> FloatValue(expr.value.floatValue)
            TokenType.STRING -> StringValue(expr.value.lexeme)
            else -> NilValue
        }
    }

    override fun visit(expr: VariableExpr) {
        try {
            lastValue = environment.get(expr.name.lexeme)
        } catch (e: UndefinedVariableException) {
            throw RuntimeException("Undefined variable: ${expr.name.lexeme}")
        }
    }

    override fun visit(expr: CallExpr) {
        val callee = evaluate(expr.callee)

        val args = mutableListOf<Value>()
        for (arg in expr.arguments) {
            args.add(evaluate(arg))
        }

        lastValue = callFunction(callee, args)
    }

    override fun visit(expr: ArrayExpr) {
        val elements = mutableListOf<Value>()
        for (elem in expr.elements) {
            elements.add(evaluate(elem))
        }
        lastValue = ArrayValue(elements)
    }

    override fun visit(expr: MemberExpr) {
        val obj = evaluate(expr.obj)

        lastValue = when (obj) {
            is ObjectValue -> obj[expr.member.lexeme]
            is DeviceValue -> {
                val dev = obj.device
                when (expr.member.lexeme) {
                    "serial" -> StringValue(dev.serial)
                    "model" -> StringValue(dev.model)
                    "screenWidth" -> IntValue(dev.screenWidth.toLong())
                    "screenHeight" -> IntValue(dev.screenHeight.toLong())
                    "androidVersion" -> StringValue(dev.androidVersion)
                    else -> throw RuntimeException("Unknown device member: ${expr.member.lexeme}")
                }
            }
            else -> throw RuntimeException("Cannot access member of non-object")
        }
    }

    override fun visit(expr: IndexExpr) {
        val obj = evaluate(expr.obj)
        val index = evaluate(expr.index)

        lastValue = when (obj) {
            is ArrayValue -> {
                if (!index.isInt()) {
                    throw RuntimeException("Array index must be an integer")
                }
                val idx = index.asInt().toInt()
                obj[idx]
            }
            is ObjectValue -> {
                if (!index.isString()) {
                    throw RuntimeException("Object key must be a string")
                }
                obj[index.asString()]
            }
            else -> throw RuntimeException("Cannot index non-array/object")
        }
    }

    // ========================================================================
    // Statement Visitors
    // ========================================================================

    override fun visit(stmt: ExpressionStmt) {
        evaluate(stmt.expression)
    }

    override fun visit(stmt: AssignmentStmt) {
        val value = evaluate(stmt.value)

        try {
            // Try to assign to existing variable
            environment.assign(stmt.variable.lexeme, value)
        } catch (e: UndefinedVariableException) {
            // If variable doesn't exist, define it
            environment.define(stmt.variable.lexeme, value)
        }
    }

    override fun visit(stmt: BlockStmt) {
        executeBlock(stmt.statements, Environment(environment))
    }

    override fun visit(stmt: IfStmt) {
        val condition = evaluate(stmt.condition)

        if (condition.isTruthy()) {
            execute(stmt.thenBranch)
        } else if (stmt.elseBranch != null) {
            execute(stmt.elseBranch)
        }
    }

    override fun visit(stmt: WhileStmt) {
        while (evaluate(stmt.condition).isTruthy()) {
            try {
                execute(stmt.body)
            } catch (e: BreakException) {
                break
            } catch (e: ContinueException) {
                continue
            }
        }
    }

    override fun visit(stmt: ForStmt) {
        // Create new scope for loop
        val loopEnv = Environment(environment)
        val previous = environment
        environment = loopEnv

        try {
            // Execute initializer
            if (stmt.initializer != null) {
                execute(stmt.initializer)
            }

            // Loop
            while (stmt.condition == null || evaluate(stmt.condition).isTruthy()) {
                try {
                    execute(stmt.body)
                } catch (e: BreakException) {
                    break
                } catch (e: ContinueException) {
                    // Continue to increment
                }

                // Execute increment
                if (stmt.increment != null) {
                    execute(stmt.increment)
                }
            }

            environment = previous
        } catch (e: Exception) {
            environment = previous
            throw e
        }
    }

    override fun visit(stmt: ForEachStmt) {
        val iterable = evaluate(stmt.iterable)

        if (iterable !is ArrayValue) {
            throw RuntimeException("ForEach requires an array")
        }

        for (item in iterable.elements) {
            // Create new scope for each iteration
            val loopEnv = Environment(environment)
            loopEnv.define(stmt.variable.lexeme, item)

            try {
                val previous = environment
                environment = loopEnv
                execute(stmt.body)
                environment = previous
            } catch (e: BreakException) {
                break
            } catch (e: ContinueException) {
                continue
            }
        }
    }

    override fun visit(stmt: FunctionStmt) {
        // Create function object
        val parameters = stmt.parameters.map { it.lexeme }
        val func = FunctionObject(parameters, stmt.body, environment)

        // Define function in environment
        environment.define(stmt.name.lexeme, FunctionValue(func))
    }

    override fun visit(stmt: ReturnStmt) {
        val value = if (stmt.value != null) {
            evaluate(stmt.value)
        } else {
            NilValue
        }
        throw ReturnException(value)
    }

    override fun visit(stmt: BreakStmt) {
        throw BreakException()
    }

    override fun visit(stmt: ContinueStmt) {
        throw ContinueException()
    }
}

// ========================================================================
// Control Flow Exceptions
// ========================================================================

class ReturnException(val value: Value) : Exception("Return")
class BreakException : Exception("Break")
class ContinueException : Exception("Continue")

File: ios-agent/iOSAgent/Runtime/Interpreter.swift
import Foundation

/**
 * Interpreter for AndroidScript
 * Executes Abstract Syntax Tree using visitor pattern
 */
class Interpreter: ASTVisitor {
    private let global: Environment
    private var environment: Environment
    private var lastValue: Value = .nil
    private var errors = [String]()

    init() {
        self.global = Environment()
        self.environment = global
    }

    // MARK: - Public Interface

    func getGlobalEnvironment() -> Environment {
        return global
    }

    func execute(statements: [Statement]) {
        for stmt in statements {
            do {
                try execute(stmt: stmt)
            } catch ControlFlow.return {
                reportError("Return statement outside of function")
            } catch ControlFlow.break {
                reportError("Break statement outside of loop")
            } catch ControlFlow.continue {
                reportError("Continue statement outside of loop")
            } catch {
                reportError("Runtime error: \(error.localizedDescription)")
            }
        }
    }

    func hasErrors() -> Bool {
        return !errors.isEmpty
    }

    func getErrors() -> [String] {
        return errors
    }

    // MARK: - Execution

    private func execute(stmt: Statement) throws {
        stmt.accept(visitor: self)
    }

    private func evaluate(expr: Expression) throws -> Value {
        expr.accept(visitor: self)
        return lastValue
    }

    private func executeBlock(statements: [Statement], environment: Environment) throws {
        let previous = self.environment
        defer { self.environment = previous }

        self.environment = environment

        for stmt in statements {
            try execute(stmt: stmt)
        }
    }

    private func callFunction(callee: Value, args: [Value]) throws -> Value {
        // Native function
        if case .nativeFunction(let function) = callee {
            return function(args)
        }

        // User-defined function
        if case .function(let function) = callee {
            // Check argument count
            guard args.count == function.parameters.count else {
                throw RuntimeError.custom("Expected \(function.parameters.count) arguments but got \(args.count)")
            }

            // Create new environment for function execution
            let funcEnv = Environment(parent: function.closure)

            // Bind parameters
            for (i, param) in function.parameters.enumerated() {
                funcEnv.define(name: param, value: args[i])
            }

            // Execute function body
            do {
                try executeBlock(statements: function.body.statements, environment: funcEnv)
                return .nil  // No explicit return
            } catch ControlFlow.return(let value) {
                return value
            }
        }

        throw RuntimeError.custom("Value is not callable")
    }

    // MARK: - Expression Visitors

    func visit(expr: BinaryExpr) {
        do {
            let left = try evaluate(expr: expr.left)
            let right = try evaluate(expr: expr.right)

            switch expr.op.type {
            case .plus:
                lastValue = try left + right
            case .minus:
                lastValue = try left - right
            case .multiply:
                lastValue = try left * right
            case .divide:
                lastValue = try left / right
            case .modulo:
                lastValue = try left % right
            case .equal:
                lastValue = .bool(left.equals(right))
            case .notEqual:
                lastValue = .bool(!left.equals(right))
            case .less:
                lastValue = .bool(try left.lessThan(right))
            case .lessEqual:
                lastValue = .bool(try left.lessEqual(right))
            case .greater:
                lastValue = .bool(try left.greaterThan(right))
            case .greaterEqual:
                lastValue = .bool(try left.greaterEqual(right))
            case .logicalAnd:
                lastValue = .bool(left.isTruthy() && right.isTruthy())
            case .logicalOr:
                lastValue = .bool(left.isTruthy() || right.isTruthy())
            default:
                throw RuntimeError.custom("Unknown binary operator: \(expr.op.type)")
            }
        } catch {
            reportError("Binary operation error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    func visit(expr: UnaryExpr) {
        do {
            let operand = try evaluate(expr: expr.operand)

            switch expr.op.type {
            case .minus:
                lastValue = try -operand
            case .logicalNot:
                lastValue = !operand
            default:
                throw RuntimeError.custom("Unknown unary operator: \(expr.op.type)")
            }
        } catch {
            reportError("Unary operation error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    func visit(expr: LiteralExpr) {
        switch expr.value.type {
        case .true:
            lastValue = .bool(true)
        case .false:
            lastValue = .bool(false)
        case .nullptr:
            lastValue = .nil
        case .integer:
            lastValue = .int(expr.value.intValue)
        case .float:
            lastValue = .float(expr.value.floatValue)
        case .string:
            lastValue = .string(expr.value.lexeme)
        default:
            lastValue = .nil
        }
    }

    func visit(expr: VariableExpr) {
        do {
            lastValue = try environment.get(name: expr.name.lexeme)
        } catch {
            reportError("Undefined variable: \(expr.name.lexeme)")
            lastValue = .nil
        }
    }

    func visit(expr: CallExpr) {
        do {
            let callee = try evaluate(expr: expr.callee)

            var args = [Value]()
            for arg in expr.arguments {
                args.append(try evaluate(expr: arg))
            }

            lastValue = try callFunction(callee: callee, args: args)
        } catch {
            reportError("Function call error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    func visit(expr: ArrayExpr) {
        do {
            var elements = [Value]()
            for elem in expr.elements {
                elements.append(try evaluate(expr: elem))
            }
            lastValue = .array(elements)
        } catch {
            reportError("Array creation error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    func visit(expr: MemberExpr) {
        do {
            let object = try evaluate(expr: expr.object)

            if case .object(let map) = object {
                lastValue = map[expr.member.lexeme] ?? .nil
            } else if case .device(let device) = object {
                // Handle device member access
                switch expr.member.lexeme {
                case "serial":
                    lastValue = .string(device.serial)
                case "model":
                    lastValue = .string(device.model)
                case "screenWidth":
                    lastValue = .int(Int64(device.screenWidth))
                case "screenHeight":
                    lastValue = .int(Int64(device.screenHeight))
                case "androidVersion":
                    lastValue = .string(device.androidVersion)
                default:
                    throw RuntimeError.custom("Unknown device member: \(expr.member.lexeme)")
                }
            } else {
                throw RuntimeError.custom("Cannot access member of non-object")
            }
        } catch {
            reportError("Member access error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    func visit(expr: IndexExpr) {
        do {
            let object = try evaluate(expr: expr.object)
            let index = try evaluate(expr: expr.index)

            if case .array(let elements) = object {
                guard case .int(let idx) = index else {
                    throw RuntimeError.custom("Array index must be an integer")
                }
                guard idx >= 0 && idx < elements.count else {
                    throw RuntimeError.arrayIndexOutOfBounds(Int(idx))
                }
                lastValue = elements[Int(idx)]
            } else if case .object(let map) = object {
                guard case .string(let key) = index else {
                    throw RuntimeError.custom("Object key must be a string")
                }
                lastValue = map[key] ?? .nil
            } else {
                throw RuntimeError.custom("Cannot index non-array/object")
            }
        } catch {
            reportError("Index access error: \(error.localizedDescription)")
            lastValue = .nil
        }
    }

    // MARK: - Statement Visitors

    func visit(stmt: ExpressionStmt) {
        do {
            _ = try evaluate(expr: stmt.expression)
        } catch {
            reportError("Expression error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: AssignmentStmt) {
        do {
            let value = try evaluate(expr: stmt.value)

            // Try to assign to existing variable
            do {
                try environment.assign(name: stmt.variable.lexeme, value: value)
            } catch RuntimeError.undefinedVariable {
                // If variable doesn't exist, define it
                environment.define(name: stmt.variable.lexeme, value: value)
            }
        } catch {
            reportError("Assignment error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: BlockStmt) {
        do {
            try executeBlock(statements: stmt.statements, environment: Environment(parent: environment))
        } catch {
            reportError("Block execution error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: IfStmt) {
        do {
            let condition = try evaluate(expr: stmt.condition)

            if condition.isTruthy() {
                try execute(stmt: stmt.thenBranch)
            } else if let elseBranch = stmt.elseBranch {
                try execute(stmt: elseBranch)
            }
        } catch {
            reportError("If statement error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: WhileStmt) {
        do {
            while try evaluate(expr: stmt.condition).isTruthy() {
                do {
                    try execute(stmt: stmt.body)
                } catch ControlFlow.break {
                    break
                } catch ControlFlow.continue {
                    continue
                }
            }
        } catch {
            reportError("While loop error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: ForStmt) {
        do {
            // Create new scope for loop
            let loopEnv = Environment(parent: environment)
            let previous = environment
            environment = loopEnv

            defer { environment = previous }

            // Execute initializer
            if let initializer = stmt.initializer {
                try execute(stmt: initializer)
            }

            // Loop
            while stmt.condition == nil || try evaluate(expr: stmt.condition!).isTruthy() {
                do {
                    try execute(stmt: stmt.body)
                } catch ControlFlow.break {
                    break
                } catch ControlFlow.continue {
                    // Continue to increment
                }

                // Execute increment
                if let increment = stmt.increment {
                    try execute(stmt: increment)
                }
            }
        } catch {
            reportError("For loop error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: ForEachStmt) {
        do {
            let iterable = try evaluate(expr: stmt.iterable)

            guard case .array(let elements) = iterable else {
                throw RuntimeError.custom("ForEach requires an array")
            }

            for item in elements {
                // Create new scope for each iteration
                let loopEnv = Environment(parent: environment)
                loopEnv.define(name: stmt.variable.lexeme, value: item)

                do {
                    try executeBlock(statements: [stmt.body], environment: loopEnv)
                } catch ControlFlow.break {
                    break
                } catch ControlFlow.continue {
                    continue
                }
            }
        } catch {
            reportError("ForEach loop error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: FunctionStmt) {
        let parameters = stmt.parameters.map { $0.lexeme }
        let function = FunctionObject(parameters: parameters, body: stmt.body, closure: environment)

        environment.define(name: stmt.name.lexeme, value: .function(function))
    }

    func visit(stmt: ReturnStmt) {
        do {
            let value: Value
            if let returnValue = stmt.value {
                value = try evaluate(expr: returnValue)
            } else {
                value = .nil
            }
            throw ControlFlow.return(value)
        } catch ControlFlow.return {
            throw ControlFlow.return(lastValue)
        } catch {
            reportError("Return statement error: \(error.localizedDescription)")
        }
    }

    func visit(stmt: BreakStmt) {
        throw ControlFlow.break
    }

    func visit(stmt: ContinueStmt) {
        throw ControlFlow.continue
    }

    // MARK: - Error Reporting

    private func reportError(_ message: String) {
        errors.append(message)
        print("Interpreter error: \(message)")
    }
}
